src/openfermion/utils/_lcu_util.py

```diff
@@ -0,0 +1,181 @@
+
+import math
+
+
+def _partial_sums(vals):
+    total = 0
+    for v in vals:
+        yield total
+        total += v
+    yield total
+
+
+def _differences(weights):
+    p = None
+    have_p = False
+    for w in weights:
+        if have_p:
+            yield w - p
+        p = w
+        have_p = True
+
+
+def _discretize_probability_distribution(unnormalized_probabilities, epsilon):
+    """Approximates probabilities with integers over a common denominator.
+
+    Args:
+        unnormalized_probabilities: A list of non-negative numbers proportional
+            to probabilities from a probability distribution. The numbers may
+            not be normalized (they do not have to add up to 1).
+        epsilon: The absolute error tolerance.
+
+    Returns:
+        A tuple containing:
+            [0]: A list of numerators.
+            [1]: A common denominator.
+
+        It is guaranteed that numerators[i] / denominator is within epsilon of
+        the i'th input probability (after normalization).
+
+        It is guaranteed that the common denominator is a multiple of the
+        number of probabilities in the distribution.
+    """
+    n = len(unnormalized_probabilities)
+    bin_count = math.ceil(1.0 / (epsilon * n)) * n
+
+    cumulative = list(_partial_sums(unnormalized_probabilities))
+    total = cumulative[-1]
+    discretized_cumulative = [int(math.floor(c/total*bin_count + 0.5))
+                              for c in cumulative]
+    discretized = list(_differences(discretized_cumulative))
+    return discretized, bin_count
+
+
+def _preprocess_for_efficient_roulette_selection(discretized_probabilities):
+    """Prepares data for performing efficient roulette selection.
+
+    The output is a tuple (alternates, keep_weights). The output is guaranteed
+    to satisfy a sampling-equivalence property. Specifically, the following
+    sampling process is guaranteed to be equivalent to simply picking index i
+    with probability weights[i] / sum(weights):
+
+        1. Pick a number i in [0, len(weights) - 1] uniformly at random.
+        2. With probability keep_weights[i] / sum(weights), return i.
+        3. Otherwise return alternates[i].
+
+    In other words, the output makes it possible to perform roulette selection
+    while generating only two random numbers, doing a single lookup of the
+    relevant (keep_chance, alternate) pair, and doing one comparison. This is
+    not so useful classically, but in the context of a quantum computation
+    where all those things are expensive the second sampling process is far
+    superior.
+
+    Args:
+        discretized_probabilities: A list of probabilities approximated by
+            integer numerators (with an implied common denominator). In order
+            to operate without floating point error, it is required that the
+            sum of this list is a multiple of the number of items in the list.
+
+    Returns:
+        A tuple containing:
```

Comment:
Same comment as above regarding how to discuss returns.

Reply:
Done.